crates/store/README.md

```diff
@@ -0,0 +1,805 @@
+
+### Database interface
+
+The `Database` trait defines the core interface for interacting with the
+underlying storage system. It provides methods for basic CRUD operations,
+iteration, and transaction application.
+
+Key operations:
+
+  - `open()`: Initialise the database with a given configuration
+  - `has()`, `get()`, `put()`, `delete()`: Basic key-value operations
+  - `iter()`: Create an iterator for range queries
+  - `apply()`: Apply a transaction to the database
+
+The `RocksDB` struct implements this trait, providing a concrete implementation
+using the [RocksDB][] storage engine.
+
+### Key structure
+
+The library uses a flexible key structure defined by the `Key` type and
+associated traits:
+
+  - `KeyComponent`: Defines a part of a key with a specific length
+  - `KeyComponents`: Combines multiple `KeyComponent`s into a complete key
+  - `AsKeyParts`: Converts a key into its column and components
+  - `FromKeyParts`: Constructs a key from its components
+
+This structure allows for efficient encoding and decoding of complex keys,
+supporting various data models.
+
+### Data types and entries
+
+The `DataType` trait defines how values are serialised and deserialised. The
+`Entry` trait combines a key with its associated data type, providing a complete
+representation of a stored item.
+
+Predefined entry types include:
+
+  - `ContextMeta`: Metadata for a context
+  - `ContextState`: State data for a context
+  - `ContextIdentity`: Identity information for a context
+  - `ContextTransaction`: Transaction data for a context
+
+### Layers
+
+The library implements a layered architecture for managing data access and
+modifications:
+
+  - `Layer`: Base trait for all layers
+  - `ReadLayer`: Provides read-only access to the underlying data
+  - `WriteLayer`: Extends `ReadLayer` with write operations
+
+### Transactions
+
+The `Transaction` struct represents a set of operations to be applied
+atomically. It supports putting and deleting entries, as well as merging with
+other transactions.
+
+### Iterators
+
+The library provides a flexible iteration system:
+
+  - `Iter`: Base iterator type
+  - `DBIter`: Trait for database-specific iterators
+
+## Operation flow
+
+  1. Initialise the store with a specific backend (e.g., [RocksDB][])
+  2. Create a `StoreHandle` to interact with the store
+  3. Perform read/write operations directly on the store
+  4. For atomic write operations:
+       - Create a `Temporal` layer
+       - Accumulate changes in the temporal layer
+       - Commit changes to the underlying store
+  5. Use transactions for atomic operations across multiple keys
+  6. Utilise iterators for range queries and scans
+
+```mermaid
+graph TD
+    A[Application] --> B[StoreHandle]
+    B --> C[Store]
+    C --> D[Database Backend]
+    D --> E[RocksDB and others]
+    B --> F[Temporal Layer]
+    F --> C
+    B --> G[Read-Only Layer]
+    G --> C
+
+subgraph Temporal Layer
+F
+H[Transaction]
+F --> H
+end
```

Comment:
what do you think of this

```mermaid
graph TD
    A[Application] --> B[StoreHandle]
    B --> C[Store]
    C --> D[Database Backend]
    D --> E[RocksDB and others]
    B --> F
    F --> C
    B --> G[Read-Only Layer]
    G --> C
subgraph F[Temporal Layer]
    H[Transaction]
end
```

<details>
<summary>patch</summary>

```suggestion
    B --> F
    F --> C
    B --> G[Read-Only Layer]
    G --> C

subgraph F[Temporal Layer]
    H[Transaction]
end
```

</details>

Reply:
Sure! If you prefer 🙂